middleware/rate_limiter.go

```diff
@@ -0,0 +1,219 @@
+package middleware
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/labstack/echo/v4"
+	"golang.org/x/time/rate"
+)
+
+type (
+	// RateLimiterStore is the interface to be implemented by custom stores.
+	RateLimiterStore interface {
+		// Stores for the rate limiter have to implement the Allow method
+		Allow(identifier string) bool
+	}
+)
+
+type (
+	// RateLimiterConfig defines the configuration for the rate limiter
+	RateLimiterConfig struct {
+		Skipper    Skipper
+		BeforeFunc BeforeFunc
+		// IdentifierExtractor uses echo.Context to extract the identifier for a visitor
+		IdentifierExtractor Extractor
+		// Store defines a store for the rate limiter
+		Store        RateLimiterStore
+		ErrorHandler ErrorHandler
+		DenyHandler  ErrorHandler
+	}
+	// ErrorHandler provides a handler for returning errors from the middleware
+	ErrorHandler func(context echo.Context) error
+	// Extractor is used to extract data from echo.Context
+	Extractor func(context echo.Context) (string, error)
+)
+
+// DefaultRateLimiterConfig defines default values for RateLimiterConfig
+var DefaultRateLimiterConfig = RateLimiterConfig{
+	Skipper: DefaultSkipper,
+	IdentifierExtractor: func(ctx echo.Context) (string, error) {
+		id := ctx.RealIP()
+		return id, nil
+	},
+	ErrorHandler: func(context echo.Context) error {
```

Comment:
> to handle an error user there should be the error to handle. For example maybe user wants to log it.
> Passing problematic identifier would be also handy - again for logging
> 
> `context echo.Context, identifier string, err error` - these 3 would be nice to have in `ErrorHandler` and `DenyHandler` signatures.

`ErrorHandler` could accept identifier and error.
For the denyhandler there is no error, so only the additional identifiert could be passed.

It could be used for futher processing in a second stage (e.g. caches too). So adding the identifier is OK for me for both.
Error only makes sense for the ErrorHandler though. 

> also. why it is not returning error? Default Echo error handler would not be able to catch it this way. It would be problematic if client is relies that `echo.HTTPErrorHandler` does logging.

We would need to make sure, to return an error that is handled by the default error handler. But in case of custom error handlers that would require users to also consider this type of error. I'm not sure about that and if that should be done now.

Reply:
by default return &echo.HTTPError as JWT middleware [does](https://github.com/labstack/echo/blob/936c48a17e97c97a54396c0243153e8621c51fa6/middleware/jwt.go#L219) and add note to errorHandler that implementor should consider/think of `echo.HTTPErrorHandler` when returning their own implementation.

for my own applications I usually create my own echo default handler to handle my app specific errors there.